src/react/branching.js

```diff
@@ -39,78 +39,74 @@ export type BranchStatusEnum = "ROOT" | "NO_BRANCH" | "NEW_BRANCH" | "BRANCH";
-    type: Value,
-    value: Value,
-  }>;
+export function getValueWithBranchingLogicApplied(realm: Realm, value: AbstractValue): Value {
+  let [condValue, consequentVal, alternateVal] = value.args;
+  invariant(condValue instanceof AbstractValue);
 
-  _applyBranchedLogicValue(realm: Realm, reactSerializerState: ReactSerializerState, value: Value): void {
-    if (
-      value instanceof StringValue ||
-      value instanceof NumberValue ||
-      value instanceof BooleanValue ||
-      value instanceof NullValue ||
-      value instanceof UndefinedValue
-    ) {
-      // terminal values
-    } else if (value instanceof ObjectValue && isReactElement(value)) {
-      addKeyToReactElement(realm, reactSerializerState, value);
-    } else if (value instanceof ArrayValue) {
-      forEachArrayValue(realm, value, elementValue => {
-        this._applyBranchedLogicValue(realm, reactSerializerState, elementValue);
-      });
-    } else if (value instanceof AbstractValue) {
-      let length = value.args.length;
-      if (length > 0) {
-        for (let i = 0; i < length; i++) {
-          this._applyBranchedLogicValue(realm, reactSerializerState, value.args[i]);
-        }
-      }
-    } else {
-      throw new ExpectedBailOut("Unsupported value encountered when applying branched logic to values");
-    }
-  }
+  const findFirstMatchingComponentTypes = (x: Value, y: Value): Value => {
+    if (x instanceof ObjectValue && isReactElement(x) && y instanceof ObjectValue && isReactElement(y)) {
+      let xType = getProperty(realm, x, "type");
+      let yType = getProperty(realm, y, "type");
 
-  applyBranchedLogic(realm: Realm, reactSerializerState: ReactSerializerState): boolean {
-    let reactElementType;
-    let applyBranchedLogic = false;
+      if (xType instanceof StringValue && yType instanceof StringValue) {
+        let xProps = getProperty(realm, x, "props");
+        let yProps = getProperty(realm, y, "props");
+        if (xProps instanceof ObjectValue && yProps instanceof ObjectValue) {
+          let xChildren = getProperty(realm, xProps, "children");
+          let yChildren = getProperty(realm, yProps, "children");
 
-    for (let i = 0; i < this._branchesToValidate.length; i++) {
-      let { type } = this._branchesToValidate[i];
-      if (reactElementType === undefined) {
-        reactElementType = type;
-      } else if (type !== reactElementType) {
-        // the types of the ReactElements do not match, so apply branch logic
-        applyBranchedLogic = true;
-        break;
-      }
-    }
-    if (applyBranchedLogic) {
-      for (let i = 0; i < this._branchesToValidate.length; i++) {
-        this._applyBranchedLogicValue(realm, reactSerializerState, this._branchesToValidate[i].value);
+          if (xChildren instanceof Value && yChildren instanceof Value) {
+            return findFirstMatchingComponentTypes(xChildren, yChildren);
+          }
+        }
+      } else if (xType === yType) {
```

Comment:
Per our discussion let's fix the bug here and add a regression test.
My sketch:

```js
var React = require('react');
// the JSX transform converts to React, so we need to add it back in
this['React'] = React;

function App(props) {
  if (props.foo) {
    return <Foo />;
  }
  return <Bar />;
}

function Foo() {
  return <input />;
}

function Bar() {
  return <input />;
}

if (this.__optimizeReactComponentTree) {
  __optimizeReactComponentTree(App);
}

module.exports = App;
```